src/wallet/silentpayment.cpp

```diff
@@ -0,0 +1,209 @@
+#include <wallet/silentpayment.h>
+
+#include <key_io.h>
+#include <secp256k1.h>
+#include <secp256k1_ecdh.h>
+#include <secp256k1_extrakeys.h>
+#include <arith_uint256.h>
+
+namespace silentpayment {
+
+Sender::Sender(const CKey sender_secret_key, const XOnlyPubKey recipient_x_only_public_key)
+{
+    m_context = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+    // Parse the recipient's XOnlyPubKey to secp256k1_xonly_pubkey
+    int return_val = secp256k1_xonly_pubkey_parse(m_context, &m_recipient_x_only_public_key, recipient_x_only_public_key.data());
+    assert(return_val);
+
+    // Create the shared secret using the recipient's x-only pubkey and the sender' secret key
+    return_val = secp256k1_ecdh_xonly(m_context, m_shared_secret, &m_recipient_x_only_public_key, sender_secret_key.begin(), nullptr, nullptr);
+    assert(return_val);
+}
+
+Sender::~Sender()
+{
+    secp256k1_context_destroy(m_context);
+    memset(m_shared_secret, 0, sizeof(m_shared_secret));
+    memset(m_recipient_x_only_public_key.data, 0, sizeof(m_recipient_x_only_public_key.data));
+}
+
+XOnlyPubKey Sender::Tweak(const int32_t& identifier)
+{
+    unsigned char shared_secret[32];
+    memcpy(shared_secret, m_shared_secret, 32);
+    assert(memcmp(shared_secret, m_shared_secret, 32) == 0);
+
+    // Add the identifier to the shared_secret
+    arith_uint256 tweak;
+    tweak = tweak + identifier;
+
+    int return_val = secp256k1_ec_seckey_tweak_add(m_context, shared_secret, ArithToUint256(tweak).data());
```

Comment:
60537b92b0f27cab91fbb61da0d195197146bff6

`return_val` should be asserted

Reply:
Done in https://github.com/bitcoin/bitcoin/pull/24897/commits/fb014fefff60d87cf21f220fa756aa449bd6daf4.